### ISSUE TITLE
Split long requests

### ISSUE THREAD

**Issue by rweigel** [hapi-server/client-python#11]
When a long time range is requested, the client needs to break the request into smaller chunks, with the chunk size consistent with https://github.com/hapi-server/data-specification/wiki/cache-specification

The branch for this is [parallel-downloads](https://github.com/hapi-server/client-python/tree/parallel-downloads).

**Comment by rweigel** [comment]
Suggested implementation to try first.

```
if opts['serial']:
    if START - STOP > 1 day:
        for i in range((START-STOP)/day):
          if log:
             log('Getting part i of N')
          data, meta = hapi(SERVER, DATASET, PARAMETERS, START + i, START + i)
          Data = append(Data, data)
        return Data, meta
else:
    # Make opts['Nparallel'] requests
    # Also add logging reporting when each part is done.
    from joblib import Parallel, delayed
    Data = Parallel(n_jobs=opts['Nparallel'])(delayed(wrapper)(i) for i in range(Ndays))
    Data = [(data,meta),(data,meta)]
    return concat(Data), Data[0][1]
```

**Comment by rweigel** [comment]
Generalizations:

* Chunk size of 1 hour, 1 day, 10 days, 1 year depending on cadence in `meta` (if reported)
* If cadence not given, make request for a small time range of data and use it to infer cadence.
* If START/STOP is not at T00:00:00, request a full day of data and trim start.

**Comment by rweigel** [comment]
* cadence < PT1S - files should contain 1 hour of data 
* PT1S <= cadence <= PT1H - files should contain 1 day of data
* cadence > PT1H - files should contain 1 month of data of data

**Comment by rweigel** [comment]
```
`parallel` (False) If `True`, make up to `n_parallel` requests to server
    in parallel (uses threads).

`n_parallel` (5) Maximum number of parallel requests to server. Max is 5.

`n_chunks` (None) Get data by making `n_chunks` requests by splitting
    requested time range.

`dt_chunk` ('infer') For requests that span a time larger than the default
    chunk size for a given dataset cadence, the client will split request
    into chunks. Allowed values of `dt_chunk` are 'infer', 'PT1M', 'PT1H',
    'P1D', and 'P1Y'. The default chunk size is determined based on the
    cadence of the dataset requested according to

        cadence < PT1S              dt_chunk='PT1H'
        PT1S <= cadence <= PT1H     dt_chunk='P1D'
        cadence > PT1H              dt_chunk='P1M'
        cadence >= P1D              dt_chunk='P1Y'

    If the dataset does not have a cadence listed in its metadata, an
    attempt is made to infer the cadence by requesting a small time range
    of data and doubling the time range until 10 records are in the response.
    The cadence used to determine the chunk size is then the average time
    difference between records.

    If requested time range is < 1/2 of chunk size, only one request is
    made. Otherwise, start and/or stop are modified to be at hour, day,
    month or year boundaries and requests are made for a time span of a
    full chunk, and trimming is performed. For example,

        Cadence = PT1M and request for
            start/stop=1999-11-12T00:10:00/stop=1999-11-12T12:09:00
            Chunk size is P1D and requested time range < 1/2 of this
            =>  Default behavior
        Cadence = PT1M and request for 
            start/stop=1999-11-12T00:10:00/1999-11-12T12:10:00
            Chunk size is P1D and requested time range >= 1/2 of this
            =>  One request with start/stop=1999-11-12/1999-11-13
                and trim performed
        Cadence = PT1M and request for
            start/stop=1999-11-12T00:10:00/1999-11-13T12:09:00
            Chunk size is P1D and requested time range > than this
            =>  Two requests:
                (1) start/stop=1999-11-12/start=1999-11-13
                (2) start/stop=1999-11-13/start=1999-11-14
                and trim performed
```

**Comment by rweigel** [comment]
@jbfaden Please comment. This is proposed documentation for the Python HAPI client that @hbatta will implement.

**Comment by jbfaden** [comment]
For the chunk size, when no cadence is specified, I would just assume a day and let the server refuse to implement with a 1408 "too much time or data requested" response if a day is too big.  The optional sample time range could be used to infer the chunk size as well.  If we really need to search for the chunk size, then I would have this as a service where it's done once.  I would hate to have to search for the chunk size again and again, and I'm sure server providers wouldn't like it either.

**Comment by rweigel** [comment]
Good point. In the logging we'll put a note that cadence is not given so that ideally the the server contact eventually gets feedback that this is missing. I'll also add to the verifier a statement about why cadence is important.

**Comment by jbfaden** [comment]
Yeah, I think we should motivate people to quote a cadence, if possible.

**Comment by rweigel** [comment]
Revision

```
If the dataset does not have a cadence listed in its metadata, an attempt is made to infer the cadence by requesting one day of data. The cadence used to determine the chunk size is then the average time difference between records.
```

**Comment by rweigel** [comment]
https://github.com/hapi-server/client-python/tree/parallel-downloads

**Comment by rweigel** [comment]
This was implemented in v0.1.7
